nn/functional/functional.go

```diff
@@ -0,0 +1,102 @@
+package functional
+
+// #cgo CFLAGS: -I ${SRCDIR}/../../cgotorch
+// #cgo LDFLAGS: -L ${SRCDIR}/../../cgotorch -Wl,-rpath ${SRCDIR}/../../cgotorch -lcgotorch
+// #cgo LDFLAGS: -L ${SRCDIR}/../../cgotorch/libtorch/lib -Wl,-rpath ${SRCDIR}/../../cgotorch/libtorch/lib -lc10 -ltorch -ltorch_cpu
+// #include "../../cgotorch/cgotorch.h"
+import "C"
+
+import (
+	"unsafe"
+
+	torch "github.com/wangkuiyi/gotorch"
+)
+
+// BatchNorm does batch nomalization for `input`
+func BatchNorm(input, weight, bias, runningMean, runningVar torch.Tensor,
+	training bool, momentum, eps float64, cudnnEnabled bool) torch.Tensor {
+	var cTraining, cCudnnEnabled C.int8_t
+	if training {
+		cTraining = 1
+	}
+	if cudnnEnabled {
+		cCudnnEnabled = 1
+
+	}
+	var cweight, cbias, cmean, cvar, t C.Tensor
+	if weight.T != nil {
+		cweight = C.Tensor(*weight.T)
+	}
+	if bias.T != nil {
+		cbias = C.Tensor(*bias.T)
+	}
+	if runningMean.T != nil {
+		cmean = C.Tensor(*runningMean.T)
+	}
+	if runningVar.T != nil {
+		cvar = C.Tensor(*runningVar.T)
+	}
+	torch.MustNil(
+		unsafe.Pointer(C.BatchNorm(
+			C.Tensor(*input.T),
+			cweight,
+			cbias,
+			cmean,
+			cvar,
+			cTraining,
+			C.double(momentum),
+			C.double(eps),
+			cCudnnEnabled,
+			&t)))
+	torch.SetTensorFinalizer((*unsafe.Pointer)(&t))
```

Comment:
Should we make SetTensorFinalizer a method of Tenosr, say Tenosr.SetFinalizer?

Reply:
The `SetTensorFinalizer ` method takes a `*C.Tensor` as input.

If we make a `Tensor. SetFinalizer ` method, we have to write codes as follows:

```go
var t C.Tensor
MustNil(unsafe.Pointer(C.RandN((*C.int64_t)(unsafe.Pointer(&shape[0])),
	C.int64_t(len(shape)), C.int64_t(rg), &t)))
tt := Tensor{(*unsafe.Pointer)(&t)}
tt.SetFinalizer()
return tt
```

The codes have 5 lines, which is longer than the current implementation(4 lines).

```go
var t C.Tensor
MustNil(unsafe.Pointer(C.RandN((*C.int64_t)(unsafe.Pointer(&shape[0])),
	C.int64_t(len(shape)), C.int64_t(rg), &t)))
SetTensorFinalizer((*unsafe.Pointer)(&t))
return Tensor{(*unsafe.Pointer)(&t)}
```